[AutoInstall] Check we have a wheel before moving the archive

gsnedders opened the pull request:
#### b04c5d174cd8818a3e9bd7b7d5b20a6f283605e7
<pre>
[AutoInstall] Check we have a wheel before moving the archive
<a href="https://bugs.webkit.org/show_bug.cgi?id=262566">https://bugs.webkit.org/show_bug.cgi?id=262566</a>

Reviewed by Jonathan Bedard.

It doesn&apos;t make sense to move other uncompressed archives into the
distribution set, we should check we&apos;re dealing with a wheel before
moving it. This makes many possible AutoInstall errors caused by our
lack of support of PEP 517 much more apparent by surfacing them at
install time.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py:
(Package.install):

Canonical link: <a href="https://commits.webkit.org/269300@main">https://commits.webkit.org/269300@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/675ea7880079e8e29835ca0f7728cf9c503fe35b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22160 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22368 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24054 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22407 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26619 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22687 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21543 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22392 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22100 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19227 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24904 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/22342 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19164 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20083 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26336 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20175 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20308 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24196 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20745 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17654 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/21964 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20097 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24307 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2767 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20693 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py:
```diff
@@ -381,7 +381,7 @@ def install(self):
                     # We might not need setup.py at all, check if we have dist-info and the library in the temporary location
                     to_be_moved = os.listdir(temp_location)
-                    if self.name not in to_be_moved and any(element.endswith('.dist-info') for element in to_be_moved):
+                    if archive.extension != 'whl' and self.name not in to_be_moved and any(element.endswith('.dist-info') for element in to_be_moved):
```
Wait, is `and` the right operator here?

Actually, are any of these operators right?! We want to reject if the archive is not a wheel _or_ if we don't have the appropriate files.

Like we only want to install if:

```python
if (
    archive.extension == "whl"
    and (self.name in to_be_moved or f"{self.name}.py" in to_be_moved)
    and f"{archive.name}-{archive.version}.dist-info" in to_be_moved
):
    # Continue installing
    pass
```

Though this code is the negation of that, so negating and applying De Morgan's law:

```python
if archive.extension != "whl" or (
    (self.name not in to_be_moved and f"{self.name}.py" not in to_be_moved)
    or f"{archive.name}-{archive.version}.dist-info" not in to_be_moved
):
    # Stop installing
    pass
```

Like, my reading of the code is that `self.name not in to_be_moved` is already meant to stop us from installing `beautifulsoup-4.12.3` rather than `bs4`?

JonWBedard replied:
@gsnedders Do we have an example of a package which currently hits this codepath and doesn't have a `whl` extension? I think all of the `and`s in this statement should actually be `or`s

gsnedders replied:
@JonWBedard Such a package would be invalid, AFAIK, and would I believe fail to install with `pip`. Unaware of anything that exists. Changing everything to `or` does require us to add the `f"{self.name}.py" not in to_be_moved` case as otherwise we will start rejecting things that are valid, I believe.

JonWBedard replied:
Conveniently, I just added the `f"{self.name}.py" not in to_be_moved` in a different change! I think changing this to be `or`, and we should land this.

gsnedders replied:
(that was #18962)

gsnedders committed:
[AutoInstall] Check we have a wheel before moving the archive
https://bugs.webkit.org/show_bug.cgi?id=262566

Reviewed by Jonathan Bedard.

It doesn't make sense to move other uncompressed archives into the
distribution set, we should check we're dealing with a wheel before
moving it. This makes many possible AutoInstall errors caused by our
lack of support of PEP 517 much more apparent by surfacing them at
install time.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py:
(Package.install):

Canonical link: https://commits.webkit.org/269300@main
